# Constitution du dataset à partir des données exportées de CEGID expert

Dossiers exportés :
- Missions du cabinet sur 2020
- Facturation 2020
- Entrées et sorties des clients

## Importation des librairies

In [1]:
import pandas as pd
import numpy as np
import re
from scipy import stats

## Lecture de l'ensemble des missions du cabinet sur 2020
1 ligne = 1 mission réalisée par 1 assistant pour le compte d'un client

In [2]:
missions = pd.read_csv("./Data/missions_2019.txt", sep="\t", decimal='.')

C:\Users\tweet\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (2,8) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [3]:
missions.head(10)

Code Mission Mission Exercice Client          Raison sociale  \
0  ACAB2019012132800     CAB     2019  09800  ABSENCES RECUPERATIONS   
1  ACOM2019009886200     COM     2019  05782           ETS MALEVILLE   
2  AJUR2019010433400     JUR     2019  A0510              TO BESCUIT   
3  AJUR2019010433400     JUR     2019  A0510              TO BESCUIT   
4  AJUR2019010457000     JUR     2019  A0477       PIVOINE & LAVANDE   
5  AJUR2019010554900     JUR     2019  A0453                  GIFRAM   
6  AJUR2019010783600     JUR     2019  03778       AUBERGE DE VIENNE   
7  AJUR2019010850600     JUR     2019  03285            CLUB SOLOGNE   
8  AJUR2019010859600     JUR     2019  03933                  E.A.L.   
9  AJUR2019010946400     JUR     2019  07651                    DTEC   

   DEPART CLIENT Assistant     ASSOCIES/EC Article  \
0            NaN       492  NON FACTURABLE     098   
1            1.0       065             JOP     049   
2            1.0       690             DAC  F06453   
3            1.0       690             DAC  F06453   
4            NaN       690             JOP  F06453   
5            NaN       690             SLE  F06453   
6            NaN       805             NAB     120   
7            NaN       805             FRL     120   
8            NaN       805             VAL     120   
9            3.0       805             FRL     120   

                                             Libellé Qté activité  \
0                             CONGES ET RECUPERATION      -112,00   
1                       AUTRES TRAVAUX EXCEPTIONNELS         0,25   
2  OUVERTURE ET PARAPHE REGISTRE LEGAL (AG, MVMTS...         1,00   
3  OUVERTURE ET PARAPHE REGISTRE LEGAL (AG, MVMTS...         1,00   
4  OUVERTURE ET PARAPHE REGISTRE LEGAL (AG, MVMTS...         1,00   
5  OUVERTURE ET PARAPHE REGISTRE LEGAL (AG, MVMTS...         1,00   
6                                              K-bis         1,00   
7                                              K-bis         1,00   
8                                              K-bis         1,00   
9                                              K-bis         1,00   

  Prix de vente Total  de vente Département            SITE  Unnamed: 15  
0         68,00       -7 616,00       SOCSJ  NON FACTURABLE          NaN  
1        155,00           38,75         ASS      PITHIVIERS          NaN  
2         35,00           35,00         INT         ORLEANS          NaN  
3         35,00           35,00         INT         ORLEANS          NaN  
4         35,00           35,00         INT      PITHIVIERS          NaN  
5         35,00           35,00         INT          OLIVET          NaN  
6          4,00            4,00        SECP         ORLEANS          NaN  
7          4,00            4,00        SECP            GIEN          NaN  
8          4,00            4,00        SECP            GIEN          NaN  
9          4,00            4,00        SECP            GIEN          NaN

In [4]:
#Le dataset des missions 2019 contient 260 722 entrées et 16 colonnes
missions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 260722 entries, 0 to 260721
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Code Mission     260722 non-null  object 
 1   Mission          260722 non-null  object 
 2   Exercice         260722 non-null  object 
 3   Client           260721 non-null  object 
 4   Raison sociale   260721 non-null  object 
 5   DEPART CLIENT    32287 non-null   float64
 6   Assistant        260721 non-null  object 
 7   ASSOCIES/EC      260721 non-null  object 
 8   Article          260721 non-null  object 
 9   Libellé          260721 non-null  object 
 10  Qté activité     260721 non-null  object 
 11  Prix de vente    260721 non-null  object 
 12  Total  de vente  260721 non-null  object 
 13  Département      258327 non-null  object 
 14  SITE             260721 non-null  object 
 15  Unnamed: 15      0 non-null       float64
dtypes: float64(2), object(14)
memory usage

In [5]:
missions.columns

Index(['Code Mission', 'Mission', 'Exercice', 'Client', 'Raison sociale',
       'DEPART CLIENT', 'Assistant', 'ASSOCIES/EC', 'Article', 'Libellé',
       'Qté activité', 'Prix de vente', 'Total  de vente', 'Département',
       'SITE', 'Unnamed: 15'],
      dtype='object')

In [6]:
#Supprimer la dernière colonne inutile
missions.drop(columns=['Département','Unnamed: 15'], inplace=True)

In [7]:
missions.columns = ['code_mission','mission','exercice','code_client',
                    'client','depart_client','code_assistant','manager','code_article',
                    'libelle_article','temps','cout_horaire','valorisation_temps','site']

In [8]:
missions.dtypes

code_mission           object
mission                object
exercice               object
code_client            object
client                 object
depart_client         float64
code_assistant         object
manager                object
code_article           object
libelle_article        object
temps                  object
cout_horaire           object
valorisation_temps     object
site                   object
dtype: object

In [9]:
missions.isna().sum()

code_mission               0
mission                    0
exercice                   0
code_client                1
client                     1
depart_client         228435
code_assistant             1
manager                    1
code_article               1
libelle_article            1
temps                      1
cout_horaire               1
valorisation_temps         1
site                       1
dtype: int64

In [10]:
#Cette ligne est à supprimer car manque trop d'information
missions[missions['code_client'].isna()].index

Int64Index([114440], dtype='int64')

In [11]:
missions.drop(114440, axis=0, inplace=True)

## Modifier les colonnes catégorielles en float

In [12]:
#Remplacer les virgules par des points et espaces pour créer un nombre décimal
missions['temps'] = [x.replace(',', '.') for x in missions['temps']]
missions['temps'] = missions['temps'].astype(float)

In [13]:
missions['cout_horaire'] = [x.replace(',', '.') for x in missions['cout_horaire']]
missions['cout_horaire'] = [re.sub('\s','',x) for x in missions['cout_horaire']]
missions['cout_horaire'] = pd.to_numeric(missions['cout_horaire'])

In [14]:
missions['valorisation_temps'] = [x.replace(',', '.') for x in missions['valorisation_temps']]
missions['valorisation_temps'] = [re.sub('\s','',x) for x in missions['valorisation_temps']]
missions['valorisation_temps'] = pd.to_numeric(missions['valorisation_temps'])

In [15]:
missions

code_mission mission exercice code_client  \
0       ACAB2019012132800     CAB     2019       09800   
1       ACOM2019009886200     COM     2019       05782   
2       AJUR2019010433400     JUR     2019       A0510   
3       AJUR2019010433400     JUR     2019       A0510   
4       AJUR2019010457000     JUR     2019       A0477   
...                   ...     ...      ...         ...   
260717  AGEJ2019012792200     GEJ     2019       06370   
260718  ACOM2019009940300     COM     2019       06665   
260719  ACOM2019009940300     COM     2019       06665   
260720  AJUR2019010972400     JUR     2019       80107   
260721  ACOM2019012822600     COM     2019       A1473   

                        client  depart_client code_assistant         manager  \
0       ABSENCES RECUPERATIONS            NaN            492  NON FACTURABLE   
1                ETS MALEVILLE            1.0            065             JOP   
2                   TO BESCUIT            1.0            690             DAC   
3                   TO BESCUIT            1.0            690             DAC   
4            PIVOINE & LAVANDE            NaN            690             JOP   
...                        ...            ...            ...             ...   
260717            NAEL CONCEPT            4.0            EPX             SLE   
260718        DE LA TERRE D'OR            NaN            292             FRL   
260719        DE LA TERRE D'OR            NaN            292             FRL   
260720        LES MAISONNETTES            NaN            335             VAL   
260721            MU SOLUTIONS            NaN            JBA             EMG   

       code_article                                    libelle_article  \
0               098                             CONGES ET RECUPERATION   
1               049                       AUTRES TRAVAUX EXCEPTIONNELS   
2            F06453  OUVERTURE ET PARAPHE REGISTRE LEGAL (AG, MVMTS...   
3            F06453  OUVERTURE ET PARAPHE REGISTRE LEGAL (AG, MVMTS...   
4            F06453  OUVERTURE ET PARAPHE REGISTRE LEGAL (AG, MVMTS...   
...             ...                                                ...   
260717           69                                   DIVERS JURIDIQUE   
260718           11                               CONTROLE ET REVISION   
260719           13          COMPTES ANNUELS  ET DECLARATIONS FISCALES   
260720           60                       SECRETARIAT JURIDIQUE ANNUEL   
260721           10                           SAISIE ET CENTRALISATION   

         temps  cout_horaire  valorisation_temps            site  
0      -112.00          68.0            -7616.00  NON FACTURABLE  
1         0.25         155.0               38.75      PITHIVIERS  
2         1.00          35.0               35.00         ORLEANS  
3         1.00          35.0               35.00         ORLEANS  
4         1.00          35.0               35.00      PITHIVIERS  
...        ...           ...                 ...             ...  
260717    0.50          71.0               35.50          OLIVET  
260718    1.00          71.0               71.00            GIEN  
260719    1.00          71.0               71.00            GIEN  
260720    0.25          66.0               16.50          BRIARE  
260721    4.75          76.5              363.38            EVRY  

[260721 rows x 14 columns]

In [16]:
missions.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 260721 entries, 0 to 260721
Data columns (total 14 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   code_mission        260721 non-null  object 
 1   mission             260721 non-null  object 
 2   exercice            260721 non-null  object 
 3   code_client         260721 non-null  object 
 4   client              260721 non-null  object 
 5   depart_client       32287 non-null   float64
 6   code_assistant      260721 non-null  object 
 7   manager             260721 non-null  object 
 8   code_article        260721 non-null  object 
 9   libelle_article     260721 non-null  object 
 10  temps               260721 non-null  float64
 11  cout_horaire        260721 non-null  float64
 12  valorisation_temps  260721 non-null  float64
 13  site                260721 non-null  object 
dtypes: float64(4), object(10)
memory usage: 29.8+ MB


## Agrégation du dataframe sur le code_mission pour sommer le temps et la valorisation du temps

Il faut garder une version non aggrégée pour avoir :
- les différents codes assistants par manager
- Les différents sites
- Les différents code_article et libellé articles (activités réalisées)

In [17]:
missions.code_mission.values

array(['ACAB2019012132800', 'ACOM2019009886200', 'AJUR2019010433400', ...,
       'ACOM2019009940300', 'AJUR2019010972400', 'ACOM2019012822600'],
      dtype=object)

In [18]:
missions_agg = missions.groupby(['code_mission']).agg({'mission':[lambda x:x.value_counts().index[0]],
                                                        'exercice':[lambda x:x.value_counts().index[0]],
                                                        'code_client': [lambda x:x.value_counts().index[0]],
                                                        'code_assistant': ['nunique'],
                                                        'manager': [lambda x:x.value_counts().index[0]],
                                                        'code_article': [lambda x: list(x.unique())],
                                                        'libelle_article': [lambda x: list(x.unique())],
                                                        'temps': ['sum'],
                                                        'valorisation_temps': ['sum'],
                                                        'site': [lambda x:x.value_counts().index[0]]})

In [19]:
missions_agg.head()

mission exercice code_client code_assistant  manager  \
                  <lambda> <lambda>    <lambda>        nunique <lambda>   
code_mission                                                              
AAGS2019012440200      AGS     2019       06924              1      JOP   
AAGS2019012498800      AGS     2019       06565              1      FRL   
AAGS2019012554600      AGS     2019       A0672              1      NAB   
ACAB2019012043400      CAB     2019       08937              6      CIT   
ACAB2019012043500      CAB     2019       08945              7      CIT   

                                           code_article  \
                                               <lambda>   
code_mission                                              
AAGS2019012440200                                 [001]   
AAGS2019012498800                                 [001]   
AAGS2019012554600                                 [001]   
ACAB2019012043400        [079, 089, 081, 078, 074, 098]   
ACAB2019012043500  [081, 098, 077, 0793, 0794, 082, 81]   

                                                     libelle_article   temps  \
                                                            <lambda>     sum   
code_mission                                                                   
AAGS2019012440200                            [RENSEIGNEMENTS CLIENT]   20.00   
AAGS2019012498800                            [RENSEIGNEMENTS CLIENT]    5.00   
AAGS2019012554600                            [RENSEIGNEMENTS CLIENT]   10.00   
ACAB2019012043400  [ORGANISATION INTERNE, ASSISTANCE INFORMATIQUE...  144.00   
ACAB2019012043500  [COMPTABILITE GENERALE CABINET, CONGES ET RECU...   12.75   

                  valorisation_temps               site  
                                 sum           <lambda>  
code_mission                                             
AAGS2019012440200            1531.00         PITHIVIERS  
AAGS2019012498800             395.50               GIEN  
AAGS2019012554600             794.00             OLIVET  
ACAB2019012043400           11051.74  DOSSIERS INTERNES  
ACAB2019012043500            1378.25  DOSSIERS INTERNES

In [20]:
missions_agg.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8878 entries, AAGS2019012440200 to ATB 2019012306800
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   (mission, <lambda>)          8878 non-null   object 
 1   (exercice, <lambda>)         8878 non-null   object 
 2   (code_client, <lambda>)      8878 non-null   object 
 3   (code_assistant, nunique)    8878 non-null   int64  
 4   (manager, <lambda>)          8878 non-null   object 
 5   (code_article, <lambda>)     8878 non-null   object 
 6   (libelle_article, <lambda>)  8878 non-null   object 
 7   (temps, sum)                 8878 non-null   float64
 8   (valorisation_temps, sum)    8878 non-null   float64
 9   (site, <lambda>)             8878 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 763.0+ KB


## Lecture du dossier facturation sur 2020
Quel coût facturé au client ?

In [21]:
facturation = pd.read_csv("./Data/facturation_2019.txt", sep="\t")

In [22]:
facturation.head()

Code mission Client  Exercice Mission  Total HT  Exercice.1  \
0  ASOP2019012859300  A1327    2019.0     SOP    109,00      2019.0   
1  ACOM2019012810200  A1328    2019.0     COM  1 129,00      2019.0   
2  ACOM2019012810900  A1332    2019.0     COM    408,00      2019.0   
3  ACOM2019012811000  A1333    2019.0     COM    427,00      2019.0   
4  ASOP2019012867900  A1334    2019.0     SOP    379,00      2019.0   

         Date Nature  N° Pièce  Article  Qté facturée Prix unitaire  \
0  20/10/2019    FAC       2.0  ACOMPTE           1.0        109,00   
1  20/10/2019    FAC       3.0  ACOMPTE           1.0      1 129,00   
2  20/10/2019    FAC       7.0  ACOMPTE           1.0        408,00   
3  20/10/2019    FAC       8.0  ACOMPTE           1.0        427,00   
4  20/10/2019    FAC       9.0  ACOMPTE           1.0        379,00   

   Numéro d'ordre                 Désignation             Raison sociale  \
0             1.0  Acompte Provisionnel N° 10      ATELIER RECTIFICATION   
1             1.0  Acompte Provisionnel N° 11   APPLICATION PRESERVATION   
2             1.0  Acompte Provisionnel N° 10               BI SOLUTIONS   
3             1.0  Acompte Provisionnel N° 10        CHARME ET SEDUCTION   
4             1.0  Acompte Provisionnel N° 10  CIRCUITS IMPRIMES THIERRY   

   Unnamed: 15  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN

In [23]:
facturation.dtypes

Code mission       object
Client             object
Exercice          float64
Mission            object
Total HT           object
Exercice.1        float64
Date               object
Nature             object
N° Pièce          float64
Article            object
Qté facturée      float64
Prix unitaire      object
Numéro d'ordre    float64
Désignation        object
Raison sociale     object
Unnamed: 15       float64
dtype: object

In [24]:
facturation[facturation['Exercice'].isna()]

Code mission          Client  Exercice  \
545  Déclarations sociales pour le 4ème trimestre  STEP NETTOYAGE       NaN   

    Mission Total HT  Exercice.1 Date Nature  N° Pièce Article  Qté facturée  \
545     NaN      NaN         NaN  NaN    NaN       NaN     NaN           NaN   

    Prix unitaire  Numéro d'ordre Désignation Raison sociale  Unnamed: 15  
545           NaN             NaN         NaN            NaN          NaN

In [25]:
#Supprimer cette ligne qui est nulle et empêche le pd.to_numeric
facturation.drop(545, axis=0, inplace=True)

In [26]:
facturation.isnull().sum()

Code mission          0
Client                0
Exercice              0
Mission               0
Total HT              0
Exercice.1            0
Date                  0
Nature                0
N° Pièce              0
Article               0
Qté facturée          0
Prix unitaire         0
Numéro d'ordre        0
Désignation           2
Raison sociale        1
Unnamed: 15       45562
dtype: int64

In [27]:
# Remplacer les valeurs catégorielles par des floats
facturation['Total HT'] = [str(x).replace(',', '.') for x in facturation['Total HT']]
facturation['Total HT'] = [re.sub('\s','',x) for x in facturation['Total HT']]
facturation['Total HT'] = pd.to_numeric(facturation['Total HT'])

In [28]:
facturation['Total HT'].isnull().sum()

0

### Remplir les valeurs NaN par 0 dans le Total HT de la facturation 

In [29]:
facturation.fillna(0, inplace=True)

In [30]:
facturation['Qté facturée'].value_counts().head(10)

1.0    36587
2.0     2417
3.0     1441
4.0     1002
5.0      713
6.0      494
7.0      462
0.0      425
8.0      297
9.0      184
Name: Qté facturée, dtype: int64

In [31]:
#Vérification que le client 00030 a bien les mêmes données que sur le logiciel
facturation.query('Client == "00030"')

Code mission Client  Exercice Mission  Total HT  Exercice.1  \
45097  ACAC2019012636400  00030    2019.0     CAC    1400.0      2019.0   
45098  ACAC2019012636400  00030    2019.0     CAC       7.0      2019.0   
45182  ACAC2019012636400  00030    2019.0     CAC    4200.0      2019.0   
45183  ACAC2019012636400  00030    2019.0     CAC      21.0      2019.0   
45346  ACAC2019012636400  00030    2019.0     CAC    1400.0      2019.0   
45347  ACAC2019012636400  00030    2019.0     CAC       7.0      2019.0   

             Date Nature  N° Pièce  Article  Qté facturée Prix unitaire  \
45097  17/01/2020    FAC     670.0  ACOMPTE           1.0      1 400,00   
45098  17/01/2020    FAC     670.0      H3C           1.0      1 400,00   
45182  30/04/2020    FAC     728.0  ACOMPTE           1.0      4 200,00   
45183  30/04/2020    FAC     728.0      H3C           1.0      4 200,00   
45346  08/07/2020    FAC     832.0  ACOMPTE           1.0      1 400,00   
45347  08/07/2020    FAC     832.0      H3C           1.0      1 400,00   

       Numéro d'ordre                Désignation               Raison sociale  \
45097             1.0  Acompte Provisionnel N° 1  ETABLISSEMENTS HENRY MILLET   
45098             3.0             Cotisation H3C  ETABLISSEMENTS HENRY MILLET   
45182             1.0  Acompte Provisionnel N° 2  ETABLISSEMENTS HENRY MILLET   
45183             2.0             Cotisation H3C  ETABLISSEMENTS HENRY MILLET   
45346             1.0  Acompte Provisionnel N° 3  ETABLISSEMENTS HENRY MILLET   
45347             3.0             Cotisation H3C  ETABLISSEMENTS HENRY MILLET   

       Unnamed: 15  
45097          0.0  
45098          0.0  
45182          0.0  
45183          0.0  
45346          0.0  
45347          0.0

## Agrégation du dataset facturation sur le code_mission pour sommer le total HT facturé au client sur une mission particulière

In [32]:
fact_agg = facturation.groupby(['Code mission']).agg({'Total HT':['sum']})

In [33]:
fact_agg.head()

Total HT
                        sum
Code mission               
AAGS2019012440200   2500.00
AAGS2019012498800   1000.00
AANT2019011972900  19509.72
AANT2019012072300   8945.42
AANT2019012087600   6831.91

Il existe plus de missions dans facturation que de missions dans le dataset mission ?

## Concaténation du dataset facturation sur l'ensemble des missions

In [34]:
data = pd.concat([missions_agg, fact_agg], axis=1)

In [35]:
data.columns = ['mission','exercice','code_client','code_assistant','manager','code_article',
                'libelle_article','temps','valorisation_temps','site','facturation']

In [36]:
#294 factures à 0 euro
data.isna().sum()

mission               407
exercice              407
code_client           407
code_assistant        407
manager               407
code_article          407
libelle_article       407
temps                 407
valorisation_temps    407
site                  407
facturation           294
dtype: int64

In [37]:
len(data.index)

9285

In [38]:
#Remplacer NaN par 0 euro
data['facturation'].fillna(0, inplace=True)

In [39]:
#Le reste des valeurs nulles correspondent au facture sur des codes_missions qui n'existent pas dans le dataset missions
data.isna().sum()

mission               407
exercice              407
code_client           407
code_assistant        407
manager               407
code_article          407
libelle_article       407
temps                 407
valorisation_temps    407
site                  407
facturation             0
dtype: int64

Il se peut que des missions ne soient pas terminées sur le millésime 2020 mais qu'il existe une facture

Supprimer les 407 lignes ajoutées sans aucune donnée car une facture mais pas dans de code_mission dans le dataset missions

In [40]:
list_index = data[data['code_client'].isna()].index
len(list_index)

407

In [41]:
data.drop(list_index, axis=0, inplace=True)

In [42]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8878 entries, AAGS2019012440200 to ATB 2019012306800
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   mission             8878 non-null   object 
 1   exercice            8878 non-null   object 
 2   code_client         8878 non-null   object 
 3   code_assistant      8878 non-null   float64
 4   manager             8878 non-null   object 
 5   code_article        8878 non-null   object 
 6   libelle_article     8878 non-null   object 
 7   temps               8878 non-null   float64
 8   valorisation_temps  8878 non-null   float64
 9   site                8878 non-null   object 
 10  facturation         8878 non-null   float64
dtypes: float64(4), object(7)
memory usage: 832.3+ KB


In [43]:
data.head()

mission exercice code_client  code_assistant manager  \
AAGS2019012440200     AGS     2019       06924             1.0     JOP   
AAGS2019012498800     AGS     2019       06565             1.0     FRL   
AAGS2019012554600     AGS     2019       A0672             1.0     NAB   
ACAB2019012043400     CAB     2019       08937             6.0     CIT   
ACAB2019012043500     CAB     2019       08945             7.0     CIT   

                                           code_article  \
AAGS2019012440200                                 [001]   
AAGS2019012498800                                 [001]   
AAGS2019012554600                                 [001]   
ACAB2019012043400        [079, 089, 081, 078, 074, 098]   
ACAB2019012043500  [081, 098, 077, 0793, 0794, 082, 81]   

                                                     libelle_article   temps  \
AAGS2019012440200                            [RENSEIGNEMENTS CLIENT]   20.00   
AAGS2019012498800                            [RENSEIGNEMENTS CLIENT]    5.00   
AAGS2019012554600                            [RENSEIGNEMENTS CLIENT]   10.00   
ACAB2019012043400  [ORGANISATION INTERNE, ASSISTANCE INFORMATIQUE...  144.00   
ACAB2019012043500  [COMPTABILITE GENERALE CABINET, CONGES ET RECU...   12.75   

                   valorisation_temps               site  facturation  
AAGS2019012440200             1531.00         PITHIVIERS       2500.0  
AAGS2019012498800              395.50               GIEN       1000.0  
AAGS2019012554600              794.00             OLIVET          0.0  
ACAB2019012043400            11051.74  DOSSIERS INTERNES      57857.6  
ACAB2019012043500             1378.25  DOSSIERS INTERNES          0.0

## Ajouter les entrées et sorties des clients MCCM

In [44]:
clients = pd.read_csv("./Data/client_dates.csv", sep=";")

In [46]:
clients_formes = pd.read_csv("./Data/client-naf-forme.csv", sep=";")

In [45]:
clients.head()

Code Date création Départ cabinet
0   30    07/11/1995              0
1   44    31/10/1995              0
2  441    30/10/1995              0
3  455    07/11/1995              0
4  549    09/11/1995              0

In [47]:
clients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11281 entries, 0 to 11280
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Code            11281 non-null  object
 1   Date création   11277 non-null  object
 2   Départ cabinet  11281 non-null  object
dtypes: object(3)
memory usage: 264.5+ KB


In [48]:
clients.dtypes

Code              object
Date création     object
Départ cabinet    object
dtype: object

In [49]:
def entree_sortie(col):
    
    for date in str(col):
        if '2019' in str(col):
            return 1
        else:
            return 0

In [50]:
#Test de la fonction
entree_sortie("20/02/2019")

1

In [51]:
clients['entrée_clt'] = clients['Date création'].apply(entree_sortie)

In [52]:
clients['sortie_clt'] = clients['Départ cabinet'].apply(entree_sortie)

In [66]:
clients.head(50)

Code Date création Départ cabinet  entrée_clt  sortie_clt
0     30    07/11/1995              0           0           0
1     44    31/10/1995              0           0           0
2    441    30/10/1995              0           0           0
3    455    07/11/1995              0           0           0
4    549    09/11/1995              0           0           0
5    569    06/11/1995              0           0           0
6    579    06/11/1995     31/03/2020           0           0
7    584    10/11/1995              0           0           0
8    613    16/11/1995              0           0           0
9    632    30/10/1995              0           0           0
10   656    06/11/1995              0           0           0
11   666    03/11/1995              0           0           0
12   676    31/10/1995              0           0           0
13   685    03/11/1995              0           0           0
14   689    15/02/1996              0           0           0
15   707    03/11/1995              0           0           0
16   713    30/10/1995     15/12/2020           0           0
17   795    31/10/1995              0           0           0
18   952    06/11/1995              0           0           0
19   961    06/11/1995              0           0           0
20  1065    31/10/1995              0           0           0
21  1078    03/11/1995              0           0           0
22  1088    03/11/1995              0           0           0
23  1169    30/10/1995              0           0           0
24  1221    31/10/1995              0           0           0
25  1259    31/10/1995              0           0           0
26  1318    30/10/1995              0           0           0
27  1349    06/11/1995              0           0           0
28  1350    31/10/1995              0           0           0
29  1353    31/10/1995              0           0           0
30  1369    10/11/1995              0           0           0
31  1395    31/10/1995              0           0           0
32  1398    06/11/1995              0           0           0
33  1412    31/10/1995              0           0           0
34  1413    10/11/1995              0           0           0
35  1420    08/01/1996              0           0           0
36  1426    31/10/1995              0           0           0
37  1440    08/11/1995              0           0           0
38  1467    06/11/1995              0           0           0
39  1476    28/02/1996              0           0           0
40  1489    03/11/1995              0           0           0
41  1498    06/11/1995              0           0           0
42  1509    17/11/1995              0           0           0
43  1534    31/10/1995              0           0           0
44  1570    31/10/1995              0           0           0
45  1590    08/11/1995              0           0           0
46  1602    03/11/1995              0           0           0
47  1626    31/10/1995              0           0           0
48  1629    06/11/1995              0           0           0
49  1650    05/04/1996              0           0           0

In [53]:
clients.columns

Index(['Code', 'Date création', 'Départ cabinet', 'entrée_clt', 'sortie_clt'], dtype='object')

In [54]:
clients.columns = ['code_client', 'ENTREE_CLIENT', 'SORTIE_CLIENT', 'entrée_clt', 'sortie_clt']

In [55]:
clients.code_client.value_counts()

A0069    1
2429     1
60202    1
11400    1
78050    1
        ..
7326     1
4994     1
A3116    1
814      1
2939     1
Name: code_client, Length: 11281, dtype: int64

## Concaténer les entrées et les sorties au dataset selon le code_client avec la fonction merge()

In [51]:
clients.code_client

0           30
1           44
2          441
3          455
4          549
         ...  
11276    P7129
11277    P7503
11278    P7540
11279    P7673
11280    P7911
Name: code_client, Length: 11281, dtype: object

In [71]:
len(clients.code_client[0])

2

In [56]:
def suppr_0(s):
    return s.lstrip("0")

In [57]:
#Test de la suppression de 0 au début des identifiants
suppr_0("06565")

'6565'

In [58]:
data.code_client = data.code_client.apply(suppr_0)

In [59]:
data.head()

mission exercice code_client  code_assistant manager  \
AAGS2019012440200     AGS     2019        6924             1.0     JOP   
AAGS2019012498800     AGS     2019        6565             1.0     FRL   
AAGS2019012554600     AGS     2019       A0672             1.0     NAB   
ACAB2019012043400     CAB     2019        8937             6.0     CIT   
ACAB2019012043500     CAB     2019        8945             7.0     CIT   

                                           code_article  \
AAGS2019012440200                                 [001]   
AAGS2019012498800                                 [001]   
AAGS2019012554600                                 [001]   
ACAB2019012043400        [079, 089, 081, 078, 074, 098]   
ACAB2019012043500  [081, 098, 077, 0793, 0794, 082, 81]   

                                                     libelle_article   temps  \
AAGS2019012440200                            [RENSEIGNEMENTS CLIENT]   20.00   
AAGS2019012498800                            [RENSEIGNEMENTS CLIENT]    5.00   
AAGS2019012554600                            [RENSEIGNEMENTS CLIENT]   10.00   
ACAB2019012043400  [ORGANISATION INTERNE, ASSISTANCE INFORMATIQUE...  144.00   
ACAB2019012043500  [COMPTABILITE GENERALE CABINET, CONGES ET RECU...   12.75   

                   valorisation_temps               site  facturation  
AAGS2019012440200             1531.00         PITHIVIERS       2500.0  
AAGS2019012498800              395.50               GIEN       1000.0  
AAGS2019012554600              794.00             OLIVET          0.0  
ACAB2019012043400            11051.74  DOSSIERS INTERNES      57857.6  
ACAB2019012043500             1378.25  DOSSIERS INTERNES          0.0

In [60]:
data[data['code_client'] == '6087SAS'].index

Index(['ACOM2019009902200'], dtype='object')

In [61]:
data['code_client'].replace(['6087SAS'], '06087SAS', inplace=True)

In [62]:
data.reset_index(inplace=True)

In [79]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8878 entries, 0 to 8877
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   index               8878 non-null   object 
 1   mission             8878 non-null   object 
 2   exercice            8878 non-null   object 
 3   code_client         8878 non-null   object 
 4   code_assistant      8878 non-null   float64
 5   manager             8878 non-null   object 
 6   code_article        8878 non-null   object 
 7   libelle_article     8878 non-null   object 
 8   temps               8878 non-null   float64
 9   valorisation_temps  8878 non-null   float64
 10  site                8878 non-null   object 
 11  facturation         8878 non-null   float64
dtypes: float64(4), object(8)
memory usage: 832.4+ KB


In [63]:
data_2019 = pd.merge(data, clients[['code_client','entrée_clt', 'sortie_clt']], how='left', on="code_client")

In [64]:
data_2019.head()

index mission exercice code_client  code_assistant manager  \
0  AAGS2019012440200     AGS     2019        6924             1.0     JOP   
1  AAGS2019012498800     AGS     2019        6565             1.0     FRL   
2  AAGS2019012554600     AGS     2019       A0672             1.0     NAB   
3  ACAB2019012043400     CAB     2019        8937             6.0     CIT   
4  ACAB2019012043500     CAB     2019        8945             7.0     CIT   

                           code_article  \
0                                 [001]   
1                                 [001]   
2                                 [001]   
3        [079, 089, 081, 078, 074, 098]   
4  [081, 098, 077, 0793, 0794, 082, 81]   

                                     libelle_article   temps  \
0                            [RENSEIGNEMENTS CLIENT]   20.00   
1                            [RENSEIGNEMENTS CLIENT]    5.00   
2                            [RENSEIGNEMENTS CLIENT]   10.00   
3  [ORGANISATION INTERNE, ASSISTANCE INFORMATIQUE...  144.00   
4  [COMPTABILITE GENERALE CABINET, CONGES ET RECU...   12.75   

   valorisation_temps               site  facturation  entrée_clt  sortie_clt  
0             1531.00         PITHIVIERS       2500.0           0           0  
1              395.50               GIEN       1000.0           0           0  
2              794.00             OLIVET          0.0           0           1  
3            11051.74  DOSSIERS INTERNES      57857.6           0           0  
4             1378.25  DOSSIERS INTERNES          0.0           0           0

In [65]:
clients_formes.head()

Code                                           Code NAF Forme juridique
0    9                                             Divers              SA
1   25                          Autre imprimerie (labeur)            SARL
2   26                   Édition de revues et périodiques              SA
3   30                  Fabrication de parquets assemblés             SAS
4   36  Commerce de détail d'habillement en magasin sp...            SARL

In [66]:
clients_formes.columns

Index(['Code', 'Code NAF', 'Forme juridique'], dtype='object')

In [67]:
clients_formes.columns = ['code_client', 'Code NAF', 'Forme juridique']

In [68]:
data_final_2019 = pd.merge(data_2019, clients_formes[['code_client', 'Code NAF','Forme juridique']], how='left', on="code_client")

In [69]:
#Vérification des valeurs nulles
data_final_2019.isna().sum()

index                 0
mission               0
exercice              0
code_client           0
code_assistant        0
manager               0
code_article          0
libelle_article       0
temps                 0
valorisation_temps    0
site                  0
facturation           0
entrée_clt            0
sortie_clt            0
Code NAF              0
Forme juridique       0
dtype: int64

In [70]:
data_final_2019.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8878 entries, 0 to 8877
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   index               8878 non-null   object 
 1   mission             8878 non-null   object 
 2   exercice            8878 non-null   object 
 3   code_client         8878 non-null   object 
 4   code_assistant      8878 non-null   float64
 5   manager             8878 non-null   object 
 6   code_article        8878 non-null   object 
 7   libelle_article     8878 non-null   object 
 8   temps               8878 non-null   float64
 9   valorisation_temps  8878 non-null   float64
 10  site                8878 non-null   object 
 11  facturation         8878 non-null   float64
 12  entrée_clt          8878 non-null   int64  
 13  sortie_clt          8878 non-null   int64  
 14  Code NAF            8878 non-null   object 
 15  Forme juridique     8878 non-null   object 
dtypes: flo

In [61]:
#Vérification de combien de lignes par client
data_final_2019.code_client.value_counts()

A0814    11
A0672     9
6565      8
4794      8
77254     8
         ..
A0522     1
50362     1
1320      1
3242      1
A1554     1
Name: code_client, Length: 4271, dtype: int64

In [63]:
data_final_2019[data_final_2019['code_client']=='7436']

index mission exercice code_client  code_assistant manager  \
2325  ACOM2019010222200     COM     2019        7436             3.0     EMG   
5577  AGS32019012308300     GS3     2019        7436             4.0     EMG   
5786  AJUR2019010691700     JUR     2019        7436             3.0     EMG   
8282  ASOP2019012254700     SOP     2019        7436            10.0     EMG   

                                           code_article  \
2325                     [010, 001, 011, 013, 019, 041]   
5577                                    [029, 024, 027]   
5786                            [069, 060, 001, F06016]   
8282  [029, 001, 020, F02083, F0217, F02171, F0200, ...   

                                        libelle_article   temps  \
2325  [SAISIE ET CENTRALISATION, RENSEIGNEMENTS CLIE...   39.25   
5577  [AUTRES TRAVAUX EN SOCIAL, FORMALITES DE LICEN...   79.50   
5786  [DIVERS JURIDIQUE, SECRETARIAT JURIDIQUE ANNUE...   15.25   
8282  [AUTRES TRAVAUX EN SOCIAL, RENSEIGNEMENTS CLIE...  224.50   

      valorisation_temps           site  facturation  entrée_clt  sortie_clt  
2325             4527.88  FONTAINEBLEAU      16600.0           0           0  
5577             6114.50  FONTAINEBLEAU       7410.0           0           0  
5786             1715.00  FONTAINEBLEAU       1200.0           0           0  
8282             9235.00  FONTAINEBLEAU       6125.0           0           0

In [74]:
#Supprimer les dossiers internes qui sont pour le cabinet et non facturable
data_final_2019[data_final_2019['type_mission'] == "SST"].index

Int64Index([8868, 8869, 8870], dtype='int64')

In [75]:
data_final_2019.drop([8868, 8869, 8870], axis=0, inplace=True)

In [76]:
data_final_2019.dtypes

code_mission           object
type_mission           object
exercice               object
code_client            object
total_assistant       float64
manager                object
code_article           object
libelle_article        object
temps                 float64
valorisation_temps    float64
site                   object
facturation           float64
entrée_clt              int64
sortie_clt              int64
secteur                object
forme                  object
dtype: object

## Exportation du dataset 2019

In [71]:
data_final_2019.columns = ['code_mission', 'type_mission', 'exercice', 'code_client', 'total_assistant',
                           'manager', 'code_article', 'libelle_article', 'temps',
                           'valorisation_temps', 'site', 'facturation', 'entrée_clt', 'sortie_clt', 'secteur', 'forme']

In [77]:
data_final_2019.head()

code_mission type_mission exercice code_client  total_assistant  \
0  AAGS2019012440200          AGS     2019        6924              1.0   
1  AAGS2019012498800          AGS     2019        6565              1.0   
2  AAGS2019012554600          AGS     2019       A0672              1.0   
3  ACAB2019012043400          CAB     2019        8937              6.0   
4  ACAB2019012043500          CAB     2019        8945              7.0   

  manager                          code_article  \
0     JOP                                 [001]   
1     FRL                                 [001]   
2     NAB                                 [001]   
3     CIT        [079, 089, 081, 078, 074, 098]   
4     CIT  [081, 098, 077, 0793, 0794, 082, 81]   

                                     libelle_article   temps  \
0                            [RENSEIGNEMENTS CLIENT]   20.00   
1                            [RENSEIGNEMENTS CLIENT]    5.00   
2                            [RENSEIGNEMENTS CLIENT]   10.00   
3  [ORGANISATION INTERNE, ASSISTANCE INFORMATIQUE...  144.00   
4  [COMPTABILITE GENERALE CABINET, CONGES ET RECU...   12.75   

   valorisation_temps               site  facturation  entrée_clt  sortie_clt  \
0             1531.00         PITHIVIERS       2500.0           0           0   
1              395.50               GIEN       1000.0           0           0   
2              794.00             OLIVET          0.0           0           1   
3            11051.74  DOSSIERS INTERNES      57857.6           0           0   
4             1378.25  DOSSIERS INTERNES          0.0           0           0   

                                             secteur forme  
0     Fabrication de meubles de bureau et de magasin   SAS  
1                           Activités d'architecture  EURL  
2  Extraction d'autres minerais de métaux non fer...   SAS  
3     Conseil en systèmes et logiciels informatiques  EURL  
4  Supports juridiques de gestion de patrimoine m...  EURL

In [78]:
data_final_2019.to_csv("./Data/dataset-2019.csv", index=False)